src/extras/core/ShapePath.js

```diff
@@ -58,230 +60,442 @@ class ShapePath {
+
+				if ( stack.length > 0 && stack[ stack.length - 1 ] === otherIntersections[ i ].identifier ) {
+
+					stack.pop();
+
+				} else {
+
+					stack.push( otherIntersections[ i ].identifier );
 
 				}
 
+				i ++;
+
 			}
 
-			if ( toChange > 0 && ambiguous === false ) {
+			stack.push( simplePath.identifier );
+
+			if ( _fillRule === 'evenodd' ) {
+
+				const isHole = stack.length % 2 === 0 ? true : false;
+				const isHoleFor = stack[ stack.length - 2 ];
+
+				return { identifier: simplePath.identifier, isHole: isHole, for: isHoleFor };
 
-				newShapeHoles = betterShapeHoles;
+			} else if ( _fillRule === 'nonzero' ) {
+
+				// check if path is a hole by counting the amount of paths with alternating rotations it has to cross.
+				let isHole = true;
+				let isHoleFor = null;
+				let lastCWValue = null;
+
+				for ( let i = 0; i < stack.length; i ++ ) {
+
+					const identifier = stack[ i ];
+					if ( isHole ) {
+
+						lastCWValue = allPaths[ identifier ].isCW;
+						isHole = false;
+						isHoleFor = identifier;
+
+					} else if ( lastCWValue !== allPaths[ identifier ].isCW ) {
+
+						lastCWValue = allPaths[ identifier ].isCW;
+						isHole = true;
+
+					}
+
+				}
+
+				return { identifier: simplePath.identifier, isHole: isHole, for: isHoleFor };
+
+			} else {
+
+				console.warn( 'fill-rule: "' + _fillRule + '" is currently not implemented.' );
 
 			}
 
 		}
 
-		let tmpHoles;
+		// check for self intersecting paths
+		// TODO
+
+		// check intersecting paths
+		// TODO
+
+		// prepare paths for hole detection
+		let identifier = 0;
+
+		let scanlineMinX = BIGNUMBER;
+		let scanlineMaxX = - BIGNUMBER;
+
+		let simplePaths = this.subPaths.map( p => {
+
+			const points = p.getPoints();
+			let maxY = - BIGNUMBER;
+			let minY = BIGNUMBER;
+			let maxX = - BIGNUMBER;
+			let minX = BIGNUMBER;
+
+	      	//points.forEach(p => p.y *= -1);
+
+			for ( let i = 0; i < points.length; i ++ ) {
+
+				const p = points[ i ];
+
+				if ( p.y > maxY ) {
 
-		for ( let i = 0, il = newShapes.length; i < il; i ++ ) {
+					maxY = p.y;
 
-			tmpShape = newShapes[ i ].s;
-			shapes.push( tmpShape );
-			tmpHoles = newShapeHoles[ i ];
+				}
+
+				if ( p.y < minY ) {
+
+					minY = p.y;
+
+				}
+
+				if ( p.x > maxX ) {
+
+					maxX = p.x;
+
+				}
+
+				if ( p.x < minX ) {
 
-			for ( let j = 0, jl = tmpHoles.length; j < jl; j ++ ) {
+					minX = p.x;
 
-				tmpShape.holes.push( tmpHoles[ j ].h );
+				}
 
 			}
 
-		}
+			//
+			if ( scanlineMaxX <= maxX ) {
+
+				scanlineMaxX = maxX + 1;
+
+			}
+
+			if ( scanlineMinX >= minX ) {
+
+				scanlineMinX = minX - 1;
+
+			}
+
+			return { curves: p.curves, points: points, isCW: ShapeUtils.isClockWise( points ), identifier: identifier ++, boundingBox: new Box2( new Vector2( minX, minY ), new Vector2( maxX, maxY ) ) };
+
+		} );
+
+		simplePaths = simplePaths.filter( sp => sp.points.length > 1 );
```

Comment:
Hello !
I don't know if this is related to the error you had in `webgl_camera_logarithmicdepthbuffer` but I tried to use this version of toShapes and for me the problems came from this filtering that was creating discrepancies in the identifiers continuity. Instead only incrementing identifiers when points.length > 0 did the trick.

Reply:
Interesting. I'll give this a try.

/ping @Ttommeke Would be great if you could review this suggestion, too. 👍

Reply:
I'll look take a look at it this afternoon.

Reply:
I made this greater than 1 instead of greater than zero to increase performance, if i'm not mistaken. A hole or shape with only one point will not be visible. Not for any technical reason. At the time I was working on svg's generated from jpeg photos, and the generator tended to add a bunch of these points to the svg.

Reply:
FYI, I am working on #24780 and there is an issue with this specific filter. It is not useless, ... but also not working correctly. 😅 Changes incoming.